File: solutions/typescript/protein-translation/1/protein-translation.ts
export function translate(rna: string = ''): string[] {
  const codonMap: { [key: string]: string } = {
    AUG: 'Methionine',
    UUU: 'Phenylalanine', UUC: 'Phenylalanine',
    UUA: 'Leucine', UUG: 'Leucine',
    UCU: 'Serine', UCC: 'Serine', UCA: 'Serine', UCG: 'Serine',
    UAU: 'Tyrosine', UAC: 'Tyrosine',
    UGU: 'Cysteine', UGC: 'Cysteine',
    UGG: 'Tryptophan',
    UAA: 'STOP', UAG: 'STOP', UGA: 'STOP'
  };
  
  const proteins: string[] = [];
  
  // Process RNA in groups of 3
  for (let i = 0; i < rna.length; i += 3) {
    const codon = rna.slice(i, i + 3);
    
    // Check if we have a complete codon
    if (codon.length !== 3) {
      throw new Error('Invalid codon');
    }
    
    const protein = codonMap[codon];
    
    // If codon is not in the map, throw error
    if (!protein) {
      throw new Error('Invalid codon');
    }
    
    // If we hit a STOP codon, stop processing
    if (protein === 'STOP') {
      break;
    }
    
    proteins.push(protein);
  }
  
  return proteins;
}

File: solutions/ruby/protein-translation/1/protein_translation.rb
class InvalidCodonError < ArgumentError; end

class Translation
  CODON_MAP = {
    'AUG' => 'Methionine',
    'UUU' => 'Phenylalanine', 'UUC' => 'Phenylalanine',
    'UUA' => 'Leucine',       'UUG' => 'Leucine',
    'UCU' => 'Serine',        'UCC' => 'Serine',        'UCA' => 'Serine', 'UCG' => 'Serine',
    'UAU' => 'Tyrosine',      'UAC' => 'Tyrosine',
    'UGU' => 'Cysteine',      'UGC' => 'Cysteine',
    'UGG' => 'Tryptophan',
    'UAA' => 'STOP',          'UAG' => 'STOP',          'UGA' => 'STOP'
  }.freeze

  def self.of_rna(strand)
    proteins = []
    stopped = false

    # Escaneamos de 3 en 3 caracteres
    strand.scan(/.../).each do |codon|
      raise InvalidCodonError, "Invalid codon: #{codon}" unless CODON_MAP.key?(codon)

      protein = CODON_MAP[codon]
      if protein == 'STOP'
        stopped = true
        break
      end

      proteins << protein
    end

    # Si la cadena no es múltiplo de 3 (quedaron letras colgando)
    # y el bucle NO se detuvo por un STOP previo, la secuencia es inválida.
    if strand.length % 3 != 0 && !stopped
      raise InvalidCodonError, "Incomplete RNA sequence"
    end

    proteins
  end
end